hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/monitoring/MonitoringCache.java

```diff
@@ -0,0 +1,157 @@
+import pl.allegro.tech.hermes.api.Topic;
+import pl.allegro.tech.hermes.api.TopicName;
+import pl.allegro.tech.hermes.common.kafka.KafkaNamesMapper;
+import pl.allegro.tech.hermes.management.config.MonitoringProperties;
+import pl.allegro.tech.hermes.management.config.kafka.AdminClientFactory;
+import pl.allegro.tech.hermes.management.config.kafka.KafkaClustersProperties;
+import pl.allegro.tech.hermes.management.config.kafka.KafkaNamesMappers;
+import pl.allegro.tech.hermes.management.domain.subscription.SubscriptionService;
+import pl.allegro.tech.hermes.management.domain.topic.TopicService;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+import static pl.allegro.tech.hermes.api.TopicName.fromQualifiedName;
+
+public class MonitoringCache {
+
+    private static final Logger logger = LoggerFactory.getLogger(MonitoringCache.class);
+
+    private final KafkaClustersProperties kafkaClustersProperties;
+    private final KafkaNamesMappers kafkaNamesMappers;
+    private final MonitoringProperties monitoringProperties;
+    private final SubscriptionService subscriptionService;
+    private final TopicService topicService;
+
+    private volatile List<TopicAndSubscription> readOnlySubscriptionsWithUnassignedPartitionsCache = new ArrayList<>();
+
+    public MonitoringCache(KafkaClustersProperties kafkaClustersProperties, KafkaNamesMappers kafkaNamesMappers,
+                           MonitoringProperties monitoringProperties, SubscriptionService subscriptionService, TopicService topicService) {
+        this.kafkaClustersProperties = kafkaClustersProperties;
+        this.kafkaNamesMappers = kafkaNamesMappers;
+        this.monitoringProperties = monitoringProperties;
+        this.subscriptionService = subscriptionService;
+        this.topicService = topicService;
+        if (monitoringProperties.isEnabled()) {
+            ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+            scheduledExecutorService.scheduleAtFixedRate(this::monitorSubscriptionsPartitions, 0,
+                    monitoringProperties.getScanEvery().getSeconds(), TimeUnit.SECONDS);
+        }
+    }
+
+    public List<TopicAndSubscription> getSubscriptionsWithUnassignedPartitions() {
+        return readOnlySubscriptionsWithUnassignedPartitionsCache;
+    }
+
+    private void monitorSubscriptionsPartitions() {
+        List<List<TopicAndSubscription>> splitSubscriptions = getSplitActiveSubscriptions();
+
+        ExecutorService executorService = Executors.newFixedThreadPool(monitoringProperties.getNumberOfThreads());
```

Comment:
creating threads is quite lengthy process, you can create the executor service in constructor

Reply:
What about monitoring that has not finished in time? A new start of monitoring will cause errors as not all threads are available.